### PR TITLE
Feat/env (working on env implement)

### PR DESCRIPTION
**Pull request by miooo0o**
### brief
The primary function of `ft_env` is not complete.
I've created a primary process that converts env to a Doubly Linked List (DLL), a subfunction that converts the DLL back to a `char ** array`, and a struct to execute it.

The subfunctions for the DLL are in src/t_env and are all tested. There willn't be any memory issues.

**add new**
-  minishell.h  : (struct) t_env, t_elst
- src/built-in/ft_env.c
- src/t_env : env_add, env_del, env_init

**norm**
src/t_env, utill, built-in.
- except  "comment is invalid in this scope" / ft_env

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```shell
$ toto="somevalue"
$ MY_VAR="$toto"
$ echo $MY_VAR
somevalue
$ 
```
These functions were created assuming the above situation!

----------------------------

```c
// struct t_env
/// @brief This struct was created with a doubly linked list
/// but can work like a deque.
/// There is a difference which is that it can be inserted and deleted
/// even if it is on an intermediate node.
/// @note Designed for env, but can be used anywhere else you need
/// to insert or delete.
/// @param key where the head of envp (i.e., before '=') is stored as 'key' 
/// @param value the value of envp (i.e., after '=') is stored.
/// Both 'key' and 'value' are dynamically allocated!
typedef struct s_env
{
	char			*key;
	char			*value;
	struct s_env	*prev;
	struct s_env	*next;
}			t_env;

/// @note For convenience, unify the internal parameter names as other.
/// @param begin front node
/// @param end rear node
typedef struct s_elst
{
	t_env	*begin;
	t_env	*end;
	int		size;
}			t_elst;
```

Assume that `envp[i]` is organized like this **`KEY=VALUE\0`**

The two structures above store environment variables are divided into `key` and `value`. This list is created considering that our shell may later set the environment variables anew or modify their values.

I don't think this is strictly necessary, as I've ( and you also) already used DLLs, and some of them can be adapted from libraries I've created in the past, so I wanted to use DLLs, albeit with a bit of complexity

----------------------------

### src/built-in/ft_env.c

As already mentioned, the functions currently created are basic to complete` ft_env`. Nevertheless, I'm pulling these functions to show how I will manipulate `env` and how to implement `$` and `$?` following later work.

I believe to use the current form to make modifying `$` and `$?` a little more comfortable!

> t_elst *env_to_dll(char **envp)

This function treats `envp[i]` as **`KEY=VALUE\0`** and stores the `key` and `value` in the `node` with memory allocations for each. I created this function with the premise that each string can be modified via different commands (like `$`).

@note: The function contains a bit of a long line, I'll comment it out so you can see it in review.

> char **dll_to_envp(t_elst *lst)

This function takes a command line split into `key` and `value` and puts them back together as **`KEY=VALUE\0`** to string.
This function contains a slightly unclean line, which I’ll comment on so that you can see it in the review.

----------------------------

### src/t_env : env_add, env_del, env_init
These are the basic DLL functions, except for the ones below.

```
void env_add_or_update(t_elst *data, char *key, char *value)
```

This function is intended for situations where you need to modify part of an `envp` using `$` or add a new `envp`.

### CODE REVIEW

File: src/t_env/env_add.c [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+void	env_add_or_update(t_elst *data, char *key, char *value)
+{
+	t_env	*current;
+	t_env	*new_node;
+
+	current = data->begin;
+	while (current)
+	{
+		if (ft_strcmp(current->key, key) == 0)
```
Comment (miooo0o) [inline src/t_env/env_add.c:89]: 앗챠 여기서 쓰려고 했구나! 나중에 수정하곘습니다. 지금 보니까 FOUND도 전달력이 좋지는 않네요.
의견 있으시면 알려주셔요.

```c
if (ft_strcmp(current->key, key) == STATUS_FOUND)
```

File: src/t_env/env_add.c [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+		(*current)->next = *new_node;
+		(*new_node)->prev = *current;
+		(*new_node)->next = next;
+		next->prev = (*new_node);
+		env_updatesize(*lst, +1);
+		return (1);
+	}
+}
+
+void	env_add_or_update(t_elst *data, char *key, char *value)
```
Comment (miooo0o) [inline src/t_env/env_add.c:81]: 환경 변수에 추가되거나 업데이트 되야하는 상황을 상정하고 만들었습니다.

```
$ alias ll="ls -la"
$ export PATH=$PATH:/new/directory/path
$ export NEW_VARIABLE=value
```

File: src/util/ft_strequ.c [hunk 3]
```diff
@@ -15,7 +15,7 @@
 /// This was created by borrowing from other code due to a notation issue.
 /// @return This function returns 1 if the strings being compared are equal
 /// and 0 if they are not.
-int		ft_strequ(char const *s1, char const *s2)
+int	ft_strequ(char const *s1, char const *s2)
```
Comment (miooo0o) [inline src/util/ft_strequ.c:18]: norm : 완료!

File: src/util/ft_strcdup.c [hunk 4]
```diff
@@ -23,10 +23,9 @@
  */
 char	*ft_strcdup(const char *src, char c)
 {
-	int	len;
 	char	*new;
-	int i;
-
+	int		len;
+	int		i;
```
Comment (miooo0o) [inline src/util/ft_strcdup.c:28]: norm : 완료!

File: test.c [hunk 5]
```diff
@@ -6,53 +6,42 @@
...
-
-// 	ft_printf("\n");
-// 	sent_print(&deque->end);
-// 	ft_printf("\n");
-// 	deque_print_all(deque);
-
-// 	sent_delall(&deque->end);
-// 	deque_del(deque);
-// }
+void ft_print_env(t_env *node)
```
Comment (miooo0o) [inline test.c:18]: 테스트 함수입니다. 지우셔도 무방합니다

**Review by miooo0o — COMMENTED** [review]
리뷰 추가했습니다!
더 설명이 필요하다면 다음 미팅에서 이야기해보아요

File: src/built-in/ft_env.c [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	t_env	*node;
+
+	lst = env_initdll();
+	i = -1;
+	while (envp[++i])
+	{
+		separator = ft_strpbrk(envp[i], "="); // separator point '=' in envp[i]
+		if (separator)
+		{
+			node = env_newnode(ft_substr(envp[i], 0, separator - envp[i]), \
```
[the remark this answers is not in the document]
Comment (san-ghun) [reply src/built-in/ft_env.c:46]: 오.... 엄청납니다 ㅎㄷㄷ

File: src/t_env/env_add.c (hunk 2 above)
Comment (san-ghun) [reply src/t_env/env_add.c:81]: 감사합니다 참고할게요!

File: src/t_env/env_del.c [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+		if (list->end != NULL)
+			list->end->next = NULL;
+	}
+	if (node_to_delete->next != NULL)
+		node_to_delete->next->prev = node_to_delete->prev;
+	if (node_to_delete->prev != NULL)
+		node_to_delete->prev->next = node_to_delete->next;
+	env_reset_node(node_to_delete);
+	free(node_to_delete);
+	env_updatesize(list, -1);
```
Comment (san-ghun) [inline src/t_env/env_del.c:48]: 노드를 리셋하고 프리해주는거 뭔가 우아하네요 ㅎㅎ 하나 또 배웁니다 감사 ㅎㅎ

File: src/built-in/ft_env.c [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+	return (path);
+}
+
+char	**dll_to_envp(t_elst *lst)
+{
+	t_env	*node;
+	char	**envp;
+	int		i;
+	size_t	key_len;
+	size_t	value_len;
```
Comment (san-ghun) [inline src/built-in/ft_env.c:82]: `key_len` `value_len` 두개 변수를 사용하지 않는다고 컴파일 체크에서 징징거리네요 ㅠ 이것 두개만 제거해서 푸시 부탁드려요

**Review by san-ghun — CHANGES_REQUESTED** [review]
`built-in/ft_env.c`
`char **dll_to_envp()`에 `key_len` `value_len` 두개 변수를 사용하지 않는다고 컴파일 체크에서 징징거리네요 ㅠ 이것 두개만 제거해서 푸시 부탁드려요.

**Comment by san-ghun** [comment]
Leaving referenced issues for the PR.
#28 #30 #31 #29

**Comment by miooo0o** [comment]
Cleared unused parameters.

**Review by san-ghun — APPROVED** [review]

**Pull request merged**